ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2450,6 +2456,269 @@ prop_connection_manager_counters serverAcc (ArbDataFlow dataFlow)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             -- This state terminates the current timeout
+             -- and starts a new one
+             TerminatingSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts newState xs
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           OutboundIdleSt _             -> case to of
+             -- Timeout terminating states
+             InboundSt Duplex ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+             TerminatedSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             -- This state terminates the current timeout
+             -- and starts a new one
+             TerminatingSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts newState xs
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           TerminatingSt                -> case to of
+             -- Timeout terminating states
+             UnnegotiatedSt Inbound ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= timeWaitTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             TerminatedSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= timeWaitTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           _ -> error ("Should be a \tau state: " ++ show st)
+      where
+        errorMsg trans time' time maxDiffTime =
+          "\nAt transition: " ++ show trans ++ "\n"
+          ++ "First happened at: " ++ show time' ++ "\n"
+          ++ "Second happened at: " ++ show time ++ "\n"
+          ++ "Should only take: "
+          ++ show maxDiffTime
+          ++ ", but took:" ++ show (diffTime time time')
+    -- If we haven't seen a \tau transition state
+    verifyTimeouts Nothing ((t, TransitionTrace _ (Transition _ to)):xs) =
+        let newState = Just (to, t)
+         in case to of
+           InboundIdleSt _       -> verifyTimeouts newState xs
+           OutboundDupSt Ticking -> verifyTimeouts newState xs
+           OutboundIdleSt _      -> verifyTimeouts newState xs
+           TerminatingSt         -> verifyTimeouts newState xs
+           _                     -> verifyTimeouts Nothing xs
+
+    getTraceEvents :: forall a b. Typeable b
+                      => SimTrace a
+                      -> [(Time, b)]
```

Comment:
if we write this function to return a `Trace` not a list of events, we will not need a separate `splitConns` (which is unfortunate, because this means we duplicate critical code for some of the tests).

There's `traceSelectEventsDynamic` which filters `Dynamic` values checking their type.
All together `transitionSignal` (a few lines above) is probably slightly more complex than:
```haskell
      transitionSignal :: [[(Time, AbstractTransitionTrace SimAddr)]]
      transitionSignal = Trace.toList                      
                       . splitConns
                       . traceSelectTraceEventsDynamic
                       $ trace
```

Reply:
The `splitConns` function that works on `Trace` does not get us the Time of the event which is needed for this particular test. That's why I duplicated the code. I gave your suggestion a go and it doesn't seem to have that type (it has `[[AbstractTransition]]` )

Reply:
I think we lose the `Time` info when we go from `SimTrace` to `Trace` via `traceSelectTraceEventsDynamic`, that's why I duplicated the code to write a version that would give me the `Time` as well.

Reply:
There's a fix for that by using contra-variance of the tracer.  One needs to define

```
data WithTime a = WithTime Time a
```
And then use
```
tracerWithTime :: Tracer m (WithTime a) -> Tracer m a
tracerWithTime (Tracer f) = Tracer $ \a -> WithTime <$> getMonotonicTime <*> f a
```

Reply:
Interesting, but I am failing to see how that is going to help, `multinodeExperiment` requires `Tracer m (WithName (Name peerAddr) (AbstractTransitionTrace peerAddr))`. And we need to extract the time from `SimEvent`, at least it seems cleaner to extract it from there than from `getMonotonicTime`

Reply:
So you need to construct:
```haskell
Tracer m (WithTime (WithName (Name peerAddr) (AbstractTransition peerAddr)))
```
pass it through `tracerWithTime` and you have what you need.

How to construct it? The same:
```haskell
tr :: Tracer m (WithTime (WithName (Name peerAddr) (AbstractTransition peerAddr)))
tr = Tracer IOSim.traceM
```

Reply:
What about adding 

```hs
traceSelectTraceEventsTime
    :: (SimEventType -> Maybe b)
    -> SimTrace a
    -> Trace (SimResult a) (Time, b)
traceSelectTraceEventsTime fn = bifoldr ( \ v _acc -> Nil v )
                                        ( \ eventCtx acc
                                         -> case fn (seType eventCtx) of
                                              Nothing -> acc
                                              Just b  -> Cons (seTime eventCtx, b) acc
                                        )
                                        undefined -- it is ignored
```

to `IOSim.hs` ?

This would avoid making a convoluted change to the code and avoid duplication in this case, and I think it is a nice, justified addition to IOSim module

Reply:
Note that [here](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/MockEnvironment.hs#L491) something similar is done.

Reply:
I'd prefer to keep `io-sim` api as small as possible, the other place should also use contravariant tracer to get the time; another thing is that this also works in `IO`, not just `io-sim`.

Reply:
I am afraid your suggestion does not typecheck :/ `Tracer` performs an `m ()` action, so all we really can do is to perform `getMonotonicTime` in order to print it, not add it to `WithTime`

Reply:
```hs
λ data WithTime a = WithTime Time a
λ :t contramapM $ \a -> flip WithTime a <$> getMonotonicTime
contramapM $ \a -> WithTime <$> getMonotonicTime <*> pure a
  :: MonadMonotonicTime m => Tracer m (WithTime a) -> Tracer m a
```